packages/animations/lib/src/page_transition_switcher.dart

```diff
@@ -0,0 +1,317 @@
+/// are the same widget and the existing widget can be updated with the new
+/// parameters. To force the transition to occur, set a [Key] on each child
+/// widget that you wish to be considered unique (typically a [ValueKey] on the
+/// widget data that distinguishes this child from the others).
+///
+/// The same key can be used for a new child as was used for an already-outgoing
+/// child; the two will not be considered related. (For example, if a progress
+/// indicator with key A is first shown, then an image with key B, then another
+/// progress indicator with key A again, all in rapid succession, then the old
+/// progress indicator and the image will be fading out while a new progress
+/// indicator is fading in.)
+class PageTransitionSwitcher extends StatefulWidget {
+  /// Creates a [PageTransitionSwitcher].
+  ///
+  /// The [duration], [reverse], and [transitionBuilder] parameters
+  /// must not be null.
+  const PageTransitionSwitcher({
+    Key key,
+    this.duration = const Duration(milliseconds: 300),
+    this.reverse = false,
+    @required this.transitionBuilder,
+    this.child,
+  })  : assert(duration != null),
+        assert(reverse != null),
+        assert(transitionBuilder != null),
+        super(key: key);
+
+  /// The current child widget to display.
+  ///
+  /// If there was a previous child, it will be transitioning out using the
+  /// secondary animation of the [transitionBuilder], while the new child
+  /// transitions in using the primary animation of the [transitionBuilder].
+  ///
+  /// If there was no previous child, then this child will transition in using
+  /// the primary animation of the [transitionBuilder].
+  ///
+  /// The child is considered to be "new" if it has a different type or [Key]
+  /// (see [Widget.canUpdate]).
+  final Widget child;
+
+  /// The duration of the transition from the old [child] value to the new one.
+  ///
+  /// This duration is applied to the given [child] when that property is set to
+  /// a new child. Changing [duration] will not affect the
+  /// durations of transitions already in progress.
+  final Duration duration;
+
+  /// Indicates the direction of the animation when a new [child] is set.
+  ///
+  /// When this is false, the new child will transition in on top of the
+  /// previously set child while its primary animation and the secondary
+  /// animation of the previous child are running forward. This is similar to
+  /// the transition associated with pushing a new [PageRoute] on top of
+  /// another.
+  ///
+  /// When this is true, the new child will transition in below the
+  /// previously set child while its secondary animation and the primary
+  /// animation of the previous child are running in reverse. This is similar to
+  /// the transition associated with popping a [PageRoute] to reveal a new
+  /// [PageRoute] below it.
+  final bool reverse;
+
+  /// A function that wraps a new [child] with a primary and secondary animation
```

Comment:
The new PageTransition child could be null, by the transitionBuilder's child is guaranteed to be non-null, right?